### PR TITLE
Possibility to exclude sensitive information

### PR DESCRIPTION
**Pull request by ostap0207**

### CODE REVIEW

**Comment by willmore** [comment]
>Need to add doc.

Add what to what doc?

**Comment by willmore** [comment]
Do we need support masking object fields?

**Comment by willmore** [comment]
Do we need to support failsafe of case insensitivity wrt field names?

**Comment by willmore** [comment]
>>Need to add doc.

> Add what to what doc?

Never mind - you need to add documentation about the feature - got it.

**Comment by ostap0207** [comment]
> Do we need support masking object fields?
I think logasm works just with hash atm, so probably don't need. @SteamRiC is that right?

File: spec/preprocessors/excluder_adapter_spec.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+  context 'when actions is "exclude"' do
+    let(:action) { 'exclude' }
+
+    it 'removes the field' do
+      expect(processed_data).not_to include(:password)
+    end
+
+    it 'removes the nested field' do
+      expect(processed_data[:nested]).not_to include(:password)
```
Comment (willmore) [inline spec/preprocessors/excluder_adapter_spec.rb]: This tests first-level nesting. Tests should include deeper nesting as well.

**Comment by ostap0207** [comment]
> Do we need to support failsafe of case insensitivity wrt field names?

Not for our purposes, afaik. I think can be added as a flag later.

File: README.md [hunk 2]
```diff
@@ -61,3 +61,45 @@ logasm = Logasm.build('myApp', { stdout: nil, logstash: { host: "localhost", por
 ```
 
 When no loggers are specified, it creates a stdout logger by default.
+
+## Preprocessors
+
+Preprocessors enable to modify data, that needs to be logged, before the actual logging.
```
Comment (willmore) [inline README.md]: Proposed rewrite: "Preprocessors support modification of log messages, prior to sending of message to the configured logger(s)."

Comment (ostap0207) [inline README.md]: Ok, but would say: "Preprocessors **enabled**/**allow** modification of log messages, prior to sending of message to the configured logger(s)." Don't really understand `support` here.

Comment (willmore) [inline README.md]: "Enable" or "allow" are fine. "Support" works as well in this context. The three terms can _generally_ be used interchangeably.

Personally, in technical writing, I find "enable" better suited for settings that can be toggled (enabled/disabled).

**Comment by willmore** [comment]
+1

**Comment by willmore** [comment]
>I think logasm works just with hash atm, so probably don't need. @SteamRiC is that right?

@SteamRiC please repsond

**Comment by SteamRiC** [comment]
https://github.com/salemove/logasm/blob/master/lib/logasm/utils.rb#L15

It tries to merge whatever you give as parameters. If it fails, it just throws. But if you merge something like `logger.info "text", {name: SOME_OBJECT}` then it only logs `SOME_OBJECT.instance_values`

**Comment by ostap0207** [comment]
@willmore added whitelister

File: README.md [hunk 3]
```diff
@@ -61,3 +61,74 @@ logasm = Logasm.build('myApp', { stdout: nil, logstash: { host: "localhost", por
...
+logger.debug("Received request", input)
+```
+
+Logger output:
+
+```
+Received request {"info":{"phone":"************"}}
+```
+
+### Whitelister
```
Comment (willmore) [inline README.md]: Whitelist will do better. Whitelister just sounds weird.

File: lib/logasm/preprocessors/excluder.rb [hunk 4]
```diff
@@ -0,0 +1,52 @@
+class Logasm
+  module Preprocessors
+    class Excluder
```
Comment (willmore) [inline lib/logasm/preprocessors/excluder.rb]: Rename Blacklist to compliment Whitelist.

File: spec/preprocessors/whitelister_spec.rb [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+  context 'with whitelisted hash' do
+    let(:pointers) { ['/array'] }
+
+    it 'does not include hash' do
+      expect(processed_data).not_to include(data: {field: 'secret'})
+    end
+  end
+
+  context 'with incorrect pointer' do
```
Comment (willmore) [inline spec/preprocessors/whitelister_spec.rb]: It's not clear what you mean by wrong/incorrect pointer. Please elaborate.

Comment (ostap0207) [inline spec/preprocessors/whitelister_spec.rb]: I meant with trailing slash, updated test desc.

Comment (willmore) [inline spec/preprocessors/whitelister_spec.rb]: Does trailing slash indicate an invalid format of pointer? or an unsupported format?

Comment (ostap0207) [inline spec/preprocessors/whitelister_spec.rb]: Ok, actually trailing slash is ok, when hash key is "field/"

Comment (ostap0207) [inline spec/preprocessors/whitelister_spec.rb]: Hm, actually no, it has to be escaped. Will add in next commit.

File: lib/logasm/preprocessors/whitelist.rb [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+      def process(data)
+        process_data('', data)
+      end
+
+      private
+
+      def validate_pointer(pointer)
+        if pointer.slice(-1) == '/'
+          raise WrongPointerException.new('Pointer should not contain trailing slash')
```
Comment (willmore) [inline lib/logasm/preprocessors/whitelist.rb]: I'm still not clear on this - is the trailing / an unsupported format? or invalid format?
Should the exception be named "InvalidPointerFormatException" or "UnsupportedPointerFormatException"

Comment (ostap0207) [inline lib/logasm/preprocessors/whitelist.rb]: Yeah, by wrong I meant invalid. You mean that `WrongPointerException`  is too general?

Comment (willmore) [inline lib/logasm/preprocessors/whitelist.rb]: `Wrong` is an odd term for technical docs. Invalid is common prefix for errors/exceptions.

**Comment by willmore** [comment]
+1

File: lib/logasm/preprocessors/blacklist.rb [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+          key = field.delete(:key)
+          mem.merge(key => {action: DEFAULT_ACTION}.merge(field))
+        end
+      end
+
+      def process(data)
+        if data.is_a? Hash
+          data.inject({}) do |mem, (key, val)|
+            if (field = @fields_to_process[key.to_s])
+              self.send("#{field[:action]}_field", mem, key, val)
```
Comment (urmastalimaa) [inline lib/logasm/preprocessors/blacklist.rb]: We should validate the actions while initializing.

File: spec/preprocessors/blacklist_spec.rb [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+  context 'when actions is "mask"' do
+    let(:action) { 'mask' }
+
+    it 'masks nested field' do
+      expect(processed_data).to include_at_depth({field: '******'}, 1)
+    end
+
+    it 'masks nested in array field' do
+      expect(processed_data[:array]).not_to include({field: 'secret'})
```
Comment (urmastalimaa) [inline spec/preprocessors/blacklist_spec.rb]: this asserts that the field is not included, but not that it is masked

Comment (urmastalimaa) [inline spec/preprocessors/blacklist_spec.rb]: Or does the previous assertion check masking?
It's confusing, maybe you can reword or refactor here.

**Comment by urmastalimaa** [comment]
A few comments, otherwise solid stuff

File: spec/preprocessors/blacklist_spec.rb (hunk 8 above)
Comment (ostap0207) [inline spec/preprocessors/blacklist_spec.rb]: Yeah, it should mask.

**Comment by ostap0207** [comment]
Updated

**Comment by ostap0207** [comment]
@willmore Thinking about adding a list of removed field in case of whitelist, so instead of
```
Publishing to hub.health: 
```
you get:
```
Publishing to hub.health: removed=[/nodeId]
```
which makes it easier to identify what in the log and what has to be whitelisted(can be disabled in config). Because right now if you enable whitelist, you will get just a bunch of empty messages, like the first one. Can be in separate PR though.
@willmore, @urmastalimaa what do you think?

**Comment by willmore** [comment]
Should we just mask everything - why remove?

**Comment by ostap0207** [comment]
Yeah, true, masking is better actually.

File: README.md [hunk 9]
```diff
@@ -106,7 +106,7 @@ Received request {"info":{"phone":"************"}}
 
 ### Whitelist
 
-Removes all the fields except those that are whitelisted in the configuration using [JSON Pointer](https://tools.ietf.org/html/rfc6901).
+Masks all the fields except those that are whitelisted in the configuration using [JSON Pointer](https://tools.ietf.org/html/rfc6901).
```
Comment (willmore) [inline README.md]: Minor edit "Masks all fields except those whitelisted in the configuration using "

**Comment by willmore** [comment]
+1

**Comment by urmastalimaa** [comment]
!squash

**Pull request merged**